enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,188 @@
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew, artifacts will be packaged within an image.  Each cluster will have an index of CLI images.  Mirrored local registries will provide binaries to disconnected environments.  
+    * We will maintain an openshift-krew-index with supported artifacts. However, with `krew index add https://someother-third-party-index` 
+    we won't limit cluster-admins from adding their own index with whatever plugins they want.  That will be unsupported, though.
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to this but that also includes sha256sum files: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will provide available CLIs and plugins for a cluster.  
+Users will continue to have the option to download binaries from either the central location where artifacts are published or from the console.  This proposal adds the mechanism to install from images 
+built from the same artifacts so all three options will offer the same binaries, ensured through sha256 signatures.  
+
+For initial implementation, krew-compatible index will be in GitHub external to cluster and extract will happen through oc.  Disconnected CLI installs and krew-index served within clusters will evolve from
+this initial implementation.
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* CLI manager will not create or update the ClusterCLI Custom Resources, those will be managed by individual CLIs and their operators. 
+
+### Requirements:
+
+1. CLI author requirements:
```

Comment:
This is not a requirements section :)

Requirements are things like:

1. No new form of binary distribution or binary creation will be addressed
2. Operators should be able to easily distribute their binaries without additional headaches
3. The version of the CLI a user is offered is appropriate for the version of the operator installed on the cluster
4. CLI experiences that do not have an operator need to be distributed like an operator because of 1
5. Arbitrary plugins not delivered by operators are not important or relevant because of 1 or 2, because anyone who wants that can just download it today from the internet.

Reply:
Thanks!  I've added these requirements.  Also, I moved the `Requirements` section up in the doc.  I added another requirement, also:
`Users must continue to have the option to download binaries from either the central location where artifacts are published or from the console.`

Reply:
> CLI experiences that do not have an operator need to be distributed like an operator because of 1

@smarterclayton isn't that something to be left to the CLI author? All that CLI manager cares about is that there is a CLI CR on the cluster and it refers to an image that contains the CLI. The CLI manager should IMHO be fine with whatever that installs that CR